File: src/main/scala/org/xbery/artbeams/users/service/UserSubscriptionServiceImpl.scala
package org.xbery.artbeams.users.service

import java.time.Instant
import java.util.{Collections, UUID}

import javax.inject.Inject
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.xbery.artbeams.common.assets.domain.AssetAttributes
import org.xbery.artbeams.common.context.OperationCtx
import org.xbery.artbeams.orders.domain.{Order, OrderItem}
import org.xbery.artbeams.orders.service.OrderService
import org.xbery.artbeams.products.service.ProductService
import org.xbery.artbeams.users.domain.{EditedUser, User}

/**
  * @author Radek Beran
  */
@Service
class UserSubscriptionServiceImpl @Inject()(userService: UserService, productService: ProductService, orderService: OrderService) extends UserSubscriptionService {

  private lazy val logger = LoggerFactory.getLogger(this.getClass)

  override def subscribe(fullName: Option[String], email: String, productId: String): Unit = {
    if (email.isEmpty) {
      throw new IllegalArgumentException(s"Cannot subscribe user with empty e-mail, fullName ${fullName}, productId ${productId}")
    }
    val user = findOrRegisterUser(fullName, email)
    findOrCreateOrderItem(user.id, productId)
  }

  override def confirmConsent(fullName: Option[String], email: String, productId: String): Option[Instant] = {
    if (email.isEmpty) {
      throw new IllegalArgumentException(s"Cannot confirm consent for user with empty e-mail, fullName ${fullName}, productId ${productId}")
    }
    val user = findOrRegisterUser(fullName, email)
    findOrCreateOrderItem(user.id, productId)
    userService.confirmConsent(email)
  }

  private def findOrCreateOrderItem(userId: String, productId: String): OrderItem = {
    // Do not create order if the product was already ordered
    // (for e.g. user can refresh the page)
    orderService.findOrderItemOfUser(userId, productId) match {
      case Some(orderItem) =>
        // order of the product already exists (at least one...)
        orderItem
      case _ =>
        // order of the product does not exist yet, create one
        createOrderOfProduct(userId, productId).items.head
    }
  }

  private def findOrRegisterUser(fullName: Option[String], email: String): User = {
    userService.findByEmail(email) match {
      case Some(user) =>
        logger.debug(s"User ${user.id}/${user.login} is already registered")
        user
      case None =>
        // Register user
        registerUser(fullName, email)
    }
  }

  private def registerUser(fullName: Option[String], email: String): User = {
    if (email.isEmpty) {
      throw new IllegalArgumentException(s"Cannot register user with empty e-mail, fullName ${fullName}")
    }
    val (firstName, lastName) = User.namesFromFullName(fullName.getOrElse(""))
    val password = UUID.randomUUID().toString() + "_" + UUID.randomUUID().toString()
    val user = EditedUser(
      AssetAttributes.EmptyId,
      email, // login = email
      password,
      password,
      firstName,
      lastName,
      email,
      Collections.emptyList()
    )
    val registeredUser = userService.saveUser(user)(OperationCtx.apply(None)).getOrElse(throw new IllegalStateException(s"Error while saving new user ${email}"))
    logger.info(s"User ${registeredUser.id}/${registeredUser.login} was registered")
    registeredUser
  }

  private def createOrderOfProduct(userId: String, productId: String): Order = {
    val commonAttributes = AssetAttributes.Empty.updatedWith(userId)
    val item = OrderItem(commonAttributes, AssetAttributes.EmptyId, productId, 1, None)
    val order = Order(commonAttributes, Seq(item))
    orderService.createOrder(order)
  }
}

File: src/main/kotlin/org/xbery/artbeams/users/service/UserSubscriptionServiceImpl.kt
package org.xbery.artbeams.users.service

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.xbery.artbeams.common.assets.domain.AssetAttributes
import org.xbery.artbeams.common.context.OperationCtx
import org.xbery.artbeams.common.context.OriginStamp
import org.xbery.artbeams.common.form.validation.ChainedEmailValidator
import org.xbery.artbeams.orders.domain.OrderState
import org.xbery.artbeams.orders.service.OrderService
import org.xbery.artbeams.products.domain.Product
import org.xbery.artbeams.users.domain.EditedUser
import org.xbery.artbeams.users.domain.User
import java.time.Instant
import java.util.*

/**
 * @author Radek Beran
 */
@Service
class UserSubscriptionServiceImpl(
    private val userService: UserService,
    private val orderService: OrderService
) : UserSubscriptionService {
    private val logger: Logger = LoggerFactory.getLogger(this::class.java)

    override fun createOrUpdateUserWithOrderAndConsent(fullName: String?, login: String, product: Product, orderNumber: String, orderState: OrderState): User {
        if (login.isEmpty()) {
            throw IllegalArgumentException("Cannot confirm consent for user with empty login (e-mail), fullName $fullName, productId ${product.id}")
        }
        val loginNormalized = login.trim().lowercase()
        val user = findOrRegisterUser(fullName, loginNormalized)
        orderService.createOrderOfProduct(user.id, product, orderNumber, orderState)
        userService.confirmConsent(user.id, product.id)
        return user
    }

    private fun findOrRegisterUser(fullName: String?, login: String): User {
        // We have users with unique logins (that are also e-mails)
        val user = userService.findByLogin(login)
        return if (user != null) {
            logger.debug("User ${user.id}/${user.login} is already registered")
            user
        } else {
            registerUser(fullName, login)
        }
    }

    private fun registerUser(fullName: String?, login: String): User {
        if (login.isEmpty()) {
            throw IllegalArgumentException("Cannot register user with empty login (e-mail), fullName $fullName")
        }
        // Validate user email
        if (!ChainedEmailValidator.isValidEmail(login)) {
            throw IllegalArgumentException("Cannot register user with invalid e-mail (login), fullName $fullName, login $login")
        }
        val names = User.namesFromFullName(fullName ?: "")
        // Generate/store random password that can be re-set later by the user
        val password = UUID.randomUUID().toString() + "_" + UUID.randomUUID().toString()
        val user = EditedUser(
          AssetAttributes.EMPTY_ID, login, password, password, names.first, names.second, listOf()
        )
        val ctx = OperationCtx(null, OriginStamp(Instant.now(), "RegisterUser", null))
        val registeredUser = userService.saveUser(user, ctx)
        logger.info("User ${registeredUser.id}/${registeredUser.login} was registered")
        return registeredUser
    }
}
